### ISSUE TITLE
Screen reader not announcing when setting breakpoints

### ISSUE THREAD

**Issue by vsccarl** [Microsoft/vscode#6491]
- VSCode Version: 1.1.0-alpha
- OS Version: Windows 10

Steps to Reproduce:

Using NVDA
1. Open a file 
2. Set a breakpoint

Nothing is being read by NVDA when hovered over or when setting a breakpoint.

**Comment by isidorn** [comment]
Hovering over should not announce anything since hover is not accessible anyways.
Setting a breakpoint is an explicit user action, and to be concise we do not repeat to the user what action he just executed.

Closing this as designed.

@zersiax can correct me if I am wrong and I can change the behavior

**Comment by zersiax** [comment]
@vsccarl let me see if I understand you correctly because I think there's two things you could be meaning here:
- You want a verbal confirmation right after you have set the breakpoint, something like 'Breakpoint set'. There's two sides of this one, @isidorn thinks to not give any representation at the moment of setting a breakpoint. While I do understand the reasoning, I'm wondering if being concise is in this case impeding user experience. Setting a breakpoint is indeed an explicit action, but how easy is it to fat-finger? E.G., how easy is it to press either the wrong hotkey for it, or pressing it by mistake?  ALso, what happens visually when you set a breakpoint, will a visual indicator be shown briefly or will it appear to the left of the line, in the margin?
- My questions have to do with the second thing you could be meaning, that is, somehow detecting  while going through the code where breakpoints have actually been set.  If there is no message about a breakpoint having been set, and there is no way to check if a line has a breakpoint on it, it is indeed an accessibility problem that a blind developer cannot see where breakpoints are in the code.  Checking for the existence of a breakpoint, if a breakpoint shows up as a visual icon next to a line, is part of a bigger issue that encompasses all visually apparent live information being unavailable to screenreaders. Think of squiggles for errors as well, for example. Would like to hear more thoughts on this one.

**Comment by isidorn** [comment]
@zersiax thanks a lot for the feedback!

When you set a breakpoint there is a visual indicator shown on the left of the line, in the margin. Due to the reasons you mentioned I am reopening this bug and we will announce once a user sets a breakpoint.

Currently a user can check the state of his breakpoints by navigating all the way to the breakpoint pane inside the debug viewlet - and he can navigate all the breakpoint items in the tree.

I agree that we should improve more on making the ux state accessible. Pinging @stevencl so he can read your comment.

**Comment by zersiax** [comment]
I thought it would be something like that. I recommend creating a new issue to report visual indicators on a per-line basis to assistive technology. Ramifications of this would be huge, major accessibility improvement if we pull this off. Proposed flow (feel free to copy to separate issue):
- use code for spawning visual indicator next to line for assistive technology logic to determine what lines to flag. Guessing medium difficulty.
- Create short, descriptive strings  to indicate the function of the indicator. 'has breakpoint' ;has error' 'has warning', 'has monkey'...whatever other indicators exist within Code. I'd be the last person to know which ones exist ;) This isn't all that difficult, but don't make them longer than 2, maybe 3 words.

Now this is where it gets interesting, because there's a number of things we can do here. What comes to mind, from most favorable to least favorable. Please chip in if anyone has other ideas:
- Append the string to the end of the line being read by the screenreader. I don't know if this is even possible since you are going to somehow have to track focus within the multiline edit field the code is in and do some ARIA magic to append the string to the end of what is being read, maybe through an invisible live region of some kind.
- Play a short sound when a particular indicator is required. A click, a short blip or ding, not entire symphonies please. We are trying to stay productive here :) Again, not sure how technically feasible this is. My JS isn't that advanced yet ;) For this approach the short strings as I suggested above would not be required
- Add a hotkey that cycles between the various visual indicators that are present, speak the type of indicator and add a third hotkey to query what for example the error contents are. Perhaps the hotkey for querying error messages is a universal concept to all these ideas, I don't know what the visual flow for this is. D you hover over the indicator and it pops up or is it more involved? The reason this is lower down the list is because we would lose the 'live' aspect. Ideally this is added in after idea 1 or 2 has already been implemented.
- Modify the line's formatting and make clear on the accessibility documentation page that screenreaders need to be set to read formatting changes. A lot of users have these settings disabled, and for good reason because it can get quite spammy. Screenreaders can query a line or character's formatting with a hotkey but only do so if they know some useful info might be gleaned from this. This will of course affect sighted users as well though, this is therefore an 'if all else fails' solution. 

What are your thoughts on all this? Implementing one of these solutions would fix this issue and a myriad of others,  while shooting Code's accessibility up into the stratosphere.

**Comment by isidorn** [comment]
Thanks a lot for feedback! For clearity , I have created a seperate issue #6583

**Comment by weinand** [comment]
Reopened: on OS X I see correct VoiceOver messages when setting and clearing a breakpoint for the first time. If I set a previously removed breakpoint again, the "Removed Breakpoint..." message stays.

**Comment by isidorn** [comment]
This is the bug with VoiceOver, they are trying to be concise by not reading the same message multiple times. However in this case the message should be re-read.
I already I correctly call aria.status. I get the same behavior if I do aria.alert so there is nothing we can do here

**Comment by weinand** [comment]
@isidorn just to be clear: I'm not expecting the same message multiple times. I'm expecting that it toggles between "Added Breakpoint..." and "Removed Breakpoint..." messages.

**Comment by isidorn** [comment]
@weinand yes, but voice over api does not support toggling. It just supports that I report a message to it. I report message A which has already been reported and for some mysterious reason(bug on their side) message A is not read out because it was read out some time ago

**Comment by zersiax** [comment]
@weinand does the message actually get spoken multiple times? As in ...does it say Removed again? Or do you come across it when navigating around?

**Comment by weinand** [comment]
@isidorn aha, now I understand. Thanks for clarification.

**Comment by zersiax** [comment]
@isidorn  I'm actually surprised you got VoiceOver to read things on OS X at all, I heard someone say they had a lot of trouble making things read on the latest OS X, where an older version apparently worked fine

**Comment by isidorn** [comment]
@zersiax it works, but there are many bugs :)

**Comment by weinand** [comment]
@zersiax I'm doing the following:

- set a breakpoint in the editor gutter
- remove the breakpoint in the editor gutter
- set breakpoint again
- remove it again
etc.

In VoiceOver I hear exactly two message:

- Breakpoint added in line...
- Breakpoint removed in line....

So I do not hear messages for the second add/remove. 

Only if I set a breakpoint in a different line and then come back to the original line, I hear messages again.
